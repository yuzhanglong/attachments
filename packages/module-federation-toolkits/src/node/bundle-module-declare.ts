import { ModuleDeclarationKind, Project, SyntaxKind } from 'ts-morph';

export interface FileOptions {
  // 声明文件路径
  path: string;

  // 声明文件模块名称
  moduleName: string;
}

/**
 * 打包类型定义文件
 *
 * @author yuzhanglong
 * @date 2021-10-03 19:28:19
 * @param fileOptions 文件相关选项，可参考上面的类型定义
 */
export const bundleModuleDeclare = (fileOptions: FileOptions[]) => {
  const project = new Project();

  const content = [];

  fileOptions.forEach(file => {
    // 添加源代码
    const source = project.addSourceFileAtPath(file.path);

    // 遍历每一个子节点，如果是 SyntaxKind.DeclareKeyword（即 declare 关键词），进行文本替换
    source.forEachDescendant(item => {
      if (item.getKind() === SyntaxKind.DeclareKeyword) {
        // 删除即可, 需要判断是不是第一个节点，否则会报异常
        item.replaceWithText(item.isFirstNodeOnLine() ? 'export' : '');
      }
    });

    // 备份根节点
    const baseStatements = source.getStructure().statements;

    // 移除现存的所有节点
    source.getStatements().forEach(res => res.remove());

    // 创建一个 module declaration，将上面备份的根节点插入之
    source.addModule({
      name: `'${file.moduleName}'`,
      declarationKind: ModuleDeclarationKind.Module,
      hasDeclareKeyword: true,
      statements: baseStatements,
    });

    // 格式化代码
    source.formatText();

    // 补充一些注释
    content.push(`// module name: ${file.moduleName}\n\n`);
    content.push(source.getText());
    content.push('\n');
  });

  return content.join('');
};
