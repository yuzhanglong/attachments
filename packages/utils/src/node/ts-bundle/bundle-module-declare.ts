import { ModuleDeclarationKind, Project, SyntaxKind } from 'ts-morph';

interface FileOptions {
  // 声明文件路径
  path: string;

  // 声明文件模块名称
  moduleName: string;
}

export const bundleModuleDeclare = (fileOptions: FileOptions[]) => {
  const project = new Project();

  const content = [];

  fileOptions.forEach(file => {
    const source = project.addSourceFileAtPath(file.path);
    source.forEachDescendant(item => {
      if (item.getKind() === SyntaxKind.DeclareKeyword) {
        // 删除即可
        item.replaceWithText('');
      }
    });

    const baseStatements = source.getStructure().statements;

    source.getStatements().forEach(res => res.remove());

    source.addModule({
      name: `'${file.moduleName}'`,
      declarationKind: ModuleDeclarationKind.Module,
      hasDeclareKeyword: true,
      statements: baseStatements,
    });

    source.formatText();
    content.push(`// module name: ${file.moduleName}\n\n`);
    content.push(source.getText());
    content.push('\n');
  });

  return content.join('');
};
