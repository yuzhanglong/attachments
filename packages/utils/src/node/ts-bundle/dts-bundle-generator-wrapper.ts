import * as os from 'os';
import * as path from 'path';
import { runCommand } from '../common/run-command';

interface BundleFileConfig {
  // 入口文件路径
  entryPath: string;

  // 输出路径
  outputPath: string;
}

/**
 * 封装 dts-bundle-generator
 *
 * @author yuzhanglong
 * @date 2021-10-01 20:09:23
 * @param entries 一个数组，详见 BundleFileConfig
 * @see BundleFileConfig
 */
export const bundleTsDeclaration = async (
  entries: BundleFileConfig[],
) => {
  // 最大并行工作数目为 cpu 核心数 - 1
  const maxWorkSize = os.cpus().length - 1;

  while (entries.length > 0) {
    const runningItems = entries.splice(0, maxWorkSize);
    await Promise.all(runningItems.map((item) => {
      const { entryPath, outputPath } = item;
      return runCommand(
        path.resolve(
          require.resolve('dts-bundle-generator'),
          '../bin/dts-bundle-generator.js',
        ), [
          entryPath,
          '--out-file',
          outputPath,
          '--project',
          path.resolve(__dirname, 'tsconfig.json'),
        ]);
    }));
  }
};
