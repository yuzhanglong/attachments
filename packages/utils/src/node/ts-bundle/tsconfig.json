{
  "compilerOptions": {
    "target": "ESNext",
    "module": "CommonJS",
    "declaration": true,
    "moduleResolution": "Node",
    "resolveJsonModule": true,
    "types": [
      "node"
    ],
    "allowJs": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "allowSyntheticDefaultImports": true,
    "sourceMap": true
  }
}
