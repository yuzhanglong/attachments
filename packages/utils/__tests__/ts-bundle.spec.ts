/**
 * File: ts-bundle.spec.ts
 * Description: typescript 类型定义打包工具封装
 * Created: 2021-10-01 14:25:20
 * Author: yuzhanglong
 * Email: yuzl1123@163.com
 */


import * as path from 'path';
import * as fs from 'fs';
import { bundleTsDeclaration } from '../src/node/ts-bundle/dts-bundle-generator-wrapper';
import { bundleModuleDeclare } from '../src/node/ts-bundle/bundle-module-declare';

describe('test ts-bundle utils', () => {
  const p = path.resolve(__dirname, 'assets');

  test('test dts bundle generator wrapper', async () => {
    await bundleTsDeclaration([
      {
        entryPath: path.resolve(p, 'bar.ts'),
        outputPath: path.resolve(p, 'bar-tmp.d.ts'),
      },
      {
        entryPath: path.resolve(p, 'foo.ts'),
        outputPath: path.resolve(p, 'foo-tmp.d.ts'),
      },
    ]);

    expect(fs.readFileSync(path.resolve(p, 'bar-tmp.d.ts')).toString()).toMatchSnapshot();
    expect(fs.readFileSync(path.resolve(p, 'foo-tmp.d.ts')).toString()).toMatchSnapshot();
  }, 20000);

  test('test bundleModuleDeclare', () => {
    const content = bundleModuleDeclare([
      {
        moduleName: 'app1/foo',
        path: path.resolve(p, 'foo-tmp.d.ts'),
      },
      {
        moduleName: 'app1/bar',
        path: path.resolve(p, 'bar-tmp.d.ts'),
      },
    ]);
    expect(content).toMatchSnapshot();
  });
});
