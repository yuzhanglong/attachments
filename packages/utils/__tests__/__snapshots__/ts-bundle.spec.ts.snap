// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ts-bundle utils test dts bundle generator wrapper 1`] = `
"// Generated by dts-bundle-generator v5.9.0

export declare const add2: (a: string, b: string) => string;
export declare class Hello2 {
	private name;
	constructor(name: string);
}
export default Hello2;

export {};
"
`;

exports[`test ts-bundle utils test dts bundle generator wrapper 2`] = `
"// Generated by dts-bundle-generator v5.9.0

export declare const add: (a: string, b: string) => string;
export declare class Hello {
	private name;
	constructor(name: string);
}
export default add;

export {};
"
`;
